components/script_injector/renderer/script_injector_render_frame_observer.h

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_SCRIPT_INJECTOR_RENDERER_SCRIPT_INJECTOR_RENDER_FRAME_OBSERVER_H_
+#define BRAVE_COMPONENTS_SCRIPT_INJECTOR_RENDERER_SCRIPT_INJECTOR_RENDER_FRAME_OBSERVER_H_
+
+#include "brave/components/script_injector/common/mojom/script_injector.mojom.h"
+
+#include <string>
+
+#include "content/public/renderer/render_frame.h"
+#include "content/public/renderer/render_frame_observer.h"
+#include "mojo/public/cpp/bindings/associated_receiver.h"
+
+namespace script_injector {
+
+class COMPONENT_EXPORT(SCRIPT_INJECTOR) ScriptInjectorRenderFrameObserver
+    : public content::RenderFrameObserver,
+      public mojom::ScriptInjector {
+ public:
+  ScriptInjectorRenderFrameObserver();
+  ScriptInjectorRenderFrameObserver(content::RenderFrame* render_frame);
+  ScriptInjectorRenderFrameObserver(const ScriptInjectorRenderFrameObserver&) =
+      delete;
+  ScriptInjectorRenderFrameObserver& operator=(
+      const ScriptInjectorRenderFrameObserver&) = delete;
+  ~ScriptInjectorRenderFrameObserver() override;
+
+  // RenderFrameObserver implementation:
+  void OnDestruct() override;
+
+  // Implement the RequestAsyncExecuteScript method.
+  COMPONENT_EXPORT(SCRIPT_INJECTOR)
+  void RequestAsyncExecuteScript(
+      int32_t world_id,
+      const std::u16string& script,
+      bool user_activation,
+      bool want_result,
+      bool await_promise,
```

Comment:
Would using the mojom types here make the caller easier to read, without changing anything about the function itself? I'm wondering because we might get callers that look like:

```c++
v.RequestAsyncExecuteScript(world_id, script, true, true, true);
```

If there's no issue with having the mojom type here, then callers like the above would be strongly typed, and would have something like:

```c++
v.RequestAsyncExecuteScript(world_id, script, blink::mojom::UserActivationOption::kActivate, blink::mojom::WantResultOption::kWantResult, blink::mojom::PromiseResultOption::kAwait);
```

This caller is much less ambiguous to the reader the than the former.

Reply:
we just copied this from 
```
void RenderFrameHostImpl::ExecuteJavaScriptForTests(
    const std::u16string& javascript,
    bool has_user_gesture,
    bool resolve_promises,
    int32_t world_id,
    JavaScriptResultAndTypeCallback callback) {
```
but what you're saying seems reasonable.

Reply:
I would normally say we should move this to a follow-up, but maybe we should change it here so it's correct for @antonok-edm to use in his PR

Reply:
Yeah, that sounds much cleaner